File: tools/online/Report.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
namespace PartyBoardOnline {
// Structured diagnostics only: never invitation, endpoint, nickname, disk path
// or arbitrary engine/exception output.
sealed class Report {
    public readonly string DirectoryPath;
    public string NativePath {get{return Path.Combine(DirectoryPath,"native.txt");}}
    readonly object gate=new object();int lines;
    static string Root {get{return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),"PartyBoard","Diagnostics");}}
    public Report(string root=null) {
        DirectoryPath=Path.Combine(root??Root,DateTime.UtcNow.ToString("yyyyMMdd-HHmmss")+"-"+Guid.NewGuid().ToString("N"));
        try{Directory.CreateDirectory(DirectoryPath);Write("format=1");}catch{}
    }
    public void Write(string text) {
        lock(gate){if(lines>=3000)return;try{File.AppendAllText(Path.Combine(DirectoryPath,"session.txt"),DateTime.UtcNow.ToString("O")+" "+text+Environment.NewLine,Encoding.UTF8);lines++;}catch{}}
    }
    // The newest desync report the game wrote. It lives with the game data,
    // not with the lobby diagnostics, and it is the only file carrying the
    // per-field values - which is what names a divergence instead of merely
    // locating it in one of sixteen subsystems.
    static string LatestDesyncReport() {
        try {
            var folder=Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                "MarioPartyRD","Party Board","netplay");
            if(!Directory.Exists(folder))return null;
            return Directory.GetFiles(folder,"netplay_desync_*.log")
                .OrderByDescending(f=>File.GetLastWriteTimeUtc(f)).FirstOrDefault();
        } catch(IOException){return null;} catch(UnauthorizedAccessException){return null;}
    }

    // Reads at most `budget` characters. `fromTail` decides which end survives
    // when the file is bigger: the running trace is worth reading backwards
    // from the moment it stopped, a desync report forwards from its header.
    static void AppendSection(StringBuilder result,string title,string path,bool fromTail,int budget) {
        if(result.Length>0 && result[result.Length-1]!=(char)10)result.AppendLine();
        result.AppendLine("--- "+title+" ---");
        if(path==null){result.AppendLine("No data available.");return;}
        try{using(var f=new FileStream(path,FileMode.Open,FileAccess.Read,FileShare.ReadWrite)){
            if(fromTail && f.Length>budget){
                f.Seek(f.Length-budget,SeekOrigin.Begin);
                result.AppendLine("[debut tronque : "+(f.Length-budget)+" octets plus anciens omis]");
            }
            using(var reader=new StreamReader(f)){
                var buffer=new char[budget];int n=reader.ReadBlock(buffer,0,buffer.Length);
                result.Append(buffer,0,n);
                if(!fromTail && reader.Peek()>=0)
                    {result.AppendLine();result.AppendLine("[fin tronquee]");}
            }
        }}catch(IOException){result.AppendLine("No data available.");}
        catch(UnauthorizedAccessException){result.AppendLine("No data available.");}
    }

    public string Read() {
        var result=new StringBuilder("PartyBoard diagnostic v1\r\n");
        AppendSection(result,"session.txt",Path.Combine(DirectoryPath,"session.txt"),false,1024*1024);
        AppendSection(result,"native.txt",Path.Combine(DirectoryPath,"native.txt"),true,1024*1024);
        AppendSection(result,"native.txt.desync",Path.Combine(DirectoryPath,"native.txt.desync"),false,256*1024);
        var desync=LatestDesyncReport();
        AppendSection(result,desync==null?"netplay_desync (absent)":Path.GetFileName(desync),
            desync,false,2*1024*1024);
        return result.ToString();
    }
    public static string ReadLatest() {
        if(!Directory.Exists(Root))throw new IOException("Créez ou rejoignez un salon avant d'exporter le diagnostic.");
        var path=Directory.GetDirectories(Root).OrderByDescending(p=>p,StringComparer.Ordinal).FirstOrDefault();
        if(path==null)throw new IOException("Aucun diagnostic disponible.");
        return new Report(path,true).Read();
    }
    Report(string path,bool existing){DirectoryPath=path;}
}
}

File: platforms/android/app/src/main/java/com/mariopartyrd/partyboard/online/Report.java
package com.mariopartyrd.partyboard.online;

import android.content.Context;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.UUID;

// tools/online/Report.cs. Structured diagnostics only: never an invitation, an
// address, a nickname, a disc path or free-form engine output.
final class Report {
    private final File directory;
    private int lines;

    private static File root(Context context) {
        return new File(context.getFilesDir(), "online-diagnostics");
    }

    private static String utc(String pattern, Date date) {
        SimpleDateFormat f = new SimpleDateFormat(pattern, Locale.ROOT);
        f.setTimeZone(TimeZone.getTimeZone("UTC"));
        return f.format(date);
    }

    Report(Context context) {
        File base = root(context);
        directory = new File(base, utc("yyyyMMdd-HHmmss", new Date()) + "-" + UUID.randomUUID().toString().replace("-", ""));
        //noinspection ResultOfMethodCallIgnored
        directory.mkdirs();
        prune(base);
        write("format=1 platform=android");
    }

    private Report(File existing) {
        directory = existing;
    }

    // Keep the ten newest sessions.
    private static void prune(File base) {
        File[] all = base.listFiles(File::isDirectory);
        if (all == null || all.length <= 10) {
            return;
        }
        Arrays.sort(all, (a, b) -> b.getName().compareTo(a.getName()));
        for (int i = 10; i < all.length; i++) {
            File[] files = all[i].listFiles();
            if (files != null) {
                for (File f : files) {
                    //noinspection ResultOfMethodCallIgnored
                    f.delete();
                }
            }
            //noinspection ResultOfMethodCallIgnored
            all[i].delete();
        }
    }

    synchronized void write(String text) {
        if (lines >= 3000) {
            return;
        }
        try (FileOutputStream out = new FileOutputStream(new File(directory, "session.txt"), true)) {
            out.write((utc("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", new Date()) + " " + text + "\n").getBytes(StandardCharsets.UTF_8));
            lines++;
        } catch (IOException ignored) {
        }
    }

    private static void section(StringBuilder result, String title, File file, int budget) {
        result.append("--- ").append(title).append(" ---\n");
        if (file == null || !file.isFile()) {
            result.append("No data available.\n");
            return;
        }
        try (RandomAccessFile f = new RandomAccessFile(file, "r")) {
            long length = f.length();
            int n = (int) Math.min(length, budget);
            byte[] bytes = new byte[n];
            f.readFully(bytes);
            result.append(new String(bytes, StandardCharsets.UTF_8));
            if (length > budget) {
                result.append("\n[fin tronquee]\n");
            }
        } catch (IOException e) {
            result.append("No data available.\n");
        }
    }

    String read() {
        StringBuilder result = new StringBuilder("PartyBoard diagnostic v1 (android)\n");
        section(result, "session.txt", new File(directory, "session.txt"), 1024 * 1024);
        return result.toString();
    }

    static String readLatest(Context context) throws IOException {
        File[] all = root(context).listFiles(File::isDirectory);
        if (all == null || all.length == 0) {
            throw new IOException(Msg.s("Créez ou rejoignez un salon avant d'exporter le diagnostic.",
                "Create or join a lobby before exporting the diagnostic."));
        }
        Arrays.sort(all, (a, b) -> b.getName().compareTo(a.getName()));
        return new Report(all[0]).read();
    }
}
